# CentOS升级Python2.7

## 更新系统和开发工具集

```
yum -y update
yum groupinstall "Development tools"
```

## 另外还需要安装 python 工具需要的额外软件包 SSL, bz2, zlib
```
 sudo yum install zlib-devel bzip2-devel openssl-devel ncurses-devel sqlite-devel readline-devel tk-devel
```

## 源码安装Python 2.7.x

```
wget http://www.python.org/ftp/python/2.7.8/Python-2.7.8.tar.xz
xz -d Python-2.7.8.tar.xz
tar -xvf Python-2.7.8.tar

# 进入目录:
cd Python-2.7.8
# 运行配置 configure:
./configure --prefix=/usr/local
# 编译安装:
make
make altinstall

# 检查 Python 版本:
[root@VM_centos ~]# python2.7 -V

Python 2.7.8

```

## 建立软连接，使系统默认的 python指向 python2.7

```
#mv /usr/bin/python /usr/bin/python2.6.6  
#ln -s /usr/local/bin/python2.7 /usr/bin/python  

# 检查
[root@VM_centos ~]# python -V
Python 2.7.8
[root@VM_centos ~]# which python 
/usr/bin/python

```

## 修复yum
> 解决系统 Python 软链接指向 Python2.7 版本后，因为yum是不兼容 Python 2.7的，所以yum不能正常工作，我们需要指定 yum 的Python版
```
#vi /usr/bin/yum  

将文件头部的
#!/usr/bin/python

改成
#!/usr/bin/python2.6.6
```

## 安装setuptools

```
#获取软件包
wget --no-check-certificate https://pypi.python.org/packages/source/s/setuptools/setuptools-1.4.2.tar.gz
# 解压:
tar -xvf setuptools-1.4.2.tar.gz
cd setuptools-1.4.2
# 使用 Python 2.7.8 安装 setuptools
python2.7 setup.py install
```

## 安装PIP

```
curl  https://bootstrap.pypa.io/get-pip.py | python2.7 -

启动pip的时候如果报错：

pkg_resources.DistributionNotFound: The 'pip==7.1.0' distribution was not found and is required by the application

检查目录：
cd /usr/local/lib/python2.7/site-packages && ls

查看目录下是否有pip-7.1.0.dist-info，如果没有把它改成pip-7.1.0.dist-info

```


## 安装pycurl报错：

> could-not-run-curl-config-errno-2-no-such-file-or-directory-when-installing

**Debian:**
```
sudo apt-get install libcurl4-openssl-dev
apt-get install libcurl4-gnutls-dev
```

**CentOS:**
```
yum install libcurl-devel
```

** python3 安装教程： http://ask.xmodulo.com/install-python3-centos.html 添加 /usrl/local/bin 到 $PATH 环境变量**

```
vim ~/.bash_profile or ~/.bashrc 
PATH=$PATH:$HOME/bin:/usr/local/bin
```


## python2.7 ERROR:root:code for hash md5 was not found

```

搜了一下，好多人在使用md5/openssl模块时，遇到以下错误，但网上提供的解决方案描述不是很清楚，这里分享下我的解决方法。

[root@svr1 backup_stript]# python
Python 2.7.11 (default, Sep  8 2016, 09:55:33) 
[GCC 4.1.2 20080704 (Red Hat 4.1.2-55)] on Linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> 
>>> import md5
ERROR:root:code for hash md5 was not found.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type md5
ERROR:root:code for hash sha1 was not found.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha1
ERROR:root:code for hash sha224 was not found.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha224
ERROR:root:code for hash sha256 was not found.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha256
ERROR:root:code for hash sha384 was not found.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha384
ERROR:root:code for hash sha512 was not found.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/local/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha512
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/md5.py", line 10, in <module>
    from hashlib import md5
ImportError: cannot import name md5


##############################################################
原因：以上错误是由于在安装python时，系统没有安装openssh模块导致的，因此我们要做的，首先是安装openssh模块

解决：
        1. 安装openssh
            yum install openssh
        2. 安装openssl(因为openssh可能会引用openssl)
            wget http://ftp.nluug.nl/security/openssl/openssl-1.1.0.tar.gz
            tar -xzvf openssl-1.1.0.tar.gz
            cd openssl-1.1.0/
            ./config --prefix=/usr/local
            make
         3. 【关键】虽然我们已经安装了相应的模块，但这些 模块后于python安装，所以，python任然无法找到对应的包。此时需要将python引用的链接库指向openssh/openssl对应的文件（这个步骤过于繁琐，这里选择直接重装python）
           tar -xzvf Python-2.7.11.tgz 
           cd Python-2.7.11/
           ./configure --prefix=/usr/local/lib
           make && make install

测试：[root@svr1 backup_stript]# python
         Python 2.7.11 (default, Sep  8 2016, 09:55:33) 
         [GCC 4.1.2 20080704 (Red Hat 4.1.2-55)] on linux2
         Type "help", "copyright", "credits" or "license" for more information.
         >>> import md5
         >>> 

问题消失。

```



## 

```

```



## 

```

```
